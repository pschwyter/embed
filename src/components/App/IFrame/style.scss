.ada-chaperone-iframe {
  outline: none;
  border: 0;
  width: 100%;
  height: 100%;
}
