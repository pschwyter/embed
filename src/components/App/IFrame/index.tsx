import Client from "models/Client";
import { Component, h } from "preact";
import { capitalize } from "services/strings";
import postMessage from "services/postMessage";
import "./style.scss";

interface InterfaceIFrame {
  client: Client,
  greetingHandle?: string,
  handle: string,
  iframeRef?: HTMLIFrameElement,
  chatURL: string,
  parentElement?: string | HTMLElement,
  styles?: string,
  metaFields?: object,
  isDrawerOpen: boolean,
  setIFrameRef(ref: HTMLIFrameElement): void,
  setIFrameLoaded(): void
}

export default class IFrame extends Component<InterfaceIFrame> {
  constructor(props: InterfaceIFrame) {
    super(props);

    this.onLoad = this.onLoad.bind(this);
    this.handleSetRef = this.handleSetRef.bind(this);
  }

  onLoad() {
    const {
      chatURL,
      parentElement,
      styles,
      greetingHandle,
      client,
      iframeRef,
      metaFields,
      setIFrameLoaded
    } = this.props;

    setIFrameLoaded();

    const followUpResponseId = client.intro && client.intro.response_id;

    const toSend = {
      ...(styles ? { styles } : {}),
      ...(metaFields ? { metaFields } : {}),
      ...(greetingHandle ? { greetingHandle } : {}),
      ...(!parentElement ? { showCloseButton: true  } : {}),
      ...(followUpResponseId ? { followUpResponseId } : {})
    };

    postMessage(iframeRef, toSend, chatURL);
  }

  handleSetRef(ref: HTMLIFrameElement) {
    const { setIFrameRef } = this.props;

    setIFrameRef(ref);
  }

  render() {
    const {
      chatURL,
      handle,
      isDrawerOpen
    } = this.props;

    return (
      <iframe
        name="ada-embed-iframe"
        className="ada-embed-iframe"
        src={chatURL}
        title={`${capitalize(handle)} Support Chat`}
        onLoad={this.onLoad}
        ref={this.handleSetRef}
        tabIndex={isDrawerOpen ? 0 : -1}
      />
    );
  }
}
