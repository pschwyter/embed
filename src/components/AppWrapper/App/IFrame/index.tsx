import { Component, h } from "preact";
import { capitalize } from "services/strings";
import postMessage from "services/postMessage";
import "./style.scss";

interface InterfaceIFrame {
  handle: string,
  iframeRef?: HTMLIFrameElement,
  chatURL: string,
  parentElement?: string | HTMLElement,
  styles?: string,
  metaFields?: object,
  isDrawerOpen: boolean,
  introShown: boolean,
  setIFrameRef(ref: HTMLIFrameElement): void,
  setIFrameLoaded(): void
}

export default class IFrame extends Component<InterfaceIFrame> {
  constructor(props: InterfaceIFrame) {
    super(props);

    this.onLoad = this.onLoad.bind(this);
    this.handleSetRef = this.handleSetRef.bind(this);
  }

  onLoad() {
    const {
      chatURL,
      parentElement,
      styles,
      iframeRef,
      setIFrameLoaded,
      introShown
    } = this.props;

    setIFrameLoaded();

    const toSend = {
      ...(styles ? { styles } : {}),
      ...(!parentElement ? { showCloseButton: true  } : {}),

      // Web interactions analytics data:
      // Indicates whether intro was shown to user
      introShown,
      // URL of the page where chat button was clicked
      initialURL: window.location.href
    };

    postMessage(iframeRef, toSend, chatURL);
  }

  handleSetRef(ref: HTMLIFrameElement) {
    const { setIFrameRef } = this.props;

    setIFrameRef(ref);
  }

  render() {
    const {
      chatURL,
      handle,
      isDrawerOpen
    } = this.props;

    return (
      <iframe
        name="ada-embed-iframe"
        className="ada-embed-iframe"
        src={chatURL}
        title={`${capitalize(handle)} Support Chat`}
        onLoad={this.onLoad}
        ref={this.handleSetRef}
        tabIndex={isDrawerOpen ? 0 : -1}
        allowFullScreen={true}
      />
    );
  }
}
