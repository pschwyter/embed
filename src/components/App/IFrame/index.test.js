/* eslint-env node, jest */

import "jsdom-global/register";
import { shallow } from "preact-render-spy";
import IFrame from "./index";

function setup(propsOverride = {}) {
  const defaultsProps = {
    client: {
      tint: "#000"
    },
    handle: "nic",
    chatURL: "https://nic.ca.ada.support/chat/",
    setIFrameRef: jest.fn()
  };

  const props = Object.assign(defaultsProps, propsOverride);
  const PRSWrapper = shallow(<IFrame {...props} />);

  return {
    props,
    PRSWrapper
  };
}

describe("<IFrame />", () => {
  it("should match snapshot", () => {
    const { PRSWrapper } = setup();

    expect(PRSWrapper).toMatchSnapshot();
  });
});
