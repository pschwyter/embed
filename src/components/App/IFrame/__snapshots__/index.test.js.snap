// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IFrame /> should match snapshot 1`] = `
preact-render-spy (1 nodes)
-------
<iframe
  name="ada-embed-iframe"
  class="ada-embed-iframe"
  src="https://nic.ca.ada.support/chat/"
  title="Nic Support Chat"
  onLoad={[Function bound ]}
  tabIndex={-1}
>
</iframe>

`;
